support for permissions via scope

bradymiller opened the pull request:
this is a re-do

1. Added several missing scopes (and adjusted the scope autofinding mechanism to find them)
2. Changed autofinding mechanism to derive the patient/ scopes from the patient-specific routes
3. Added permission mechanism in each endpoint to check for scope
4. In dispatch, ensured that had the permission (scope) for the requested route (api:oemr, api:fhir, etc.)

Issues:
1. ~~The scope validation mechanism appears to be super slow when sending lots of scopes.~~ (fixed)
2. Couple php warnings to address.

sjpadgett commented on API_README.md:
```diff
@@ -146,6 +146,40 @@ Response:
+   "client_name": "A Private App",
+   "token_endpoint_auth_method": "client_secret_post",
+   "contacts": ["me@example.org", "them@example.org"],
+   "scope": "openid user/Organization.read"
```
Have to include the fhirUser scope or a api:fhir api:pofh scope so server knows these scopes scope.

sjpadgett commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -86,6 +86,9 @@ public function getScopeEntityByIdentifier($scopeIdentifier)
             'patient/Observation.read' => [
                 'description' => 'Read only access observation resources for a patient resource.',
             ],
+            'user/Organization.read' => [
+                'description' => 'Read only access organization resources for a patient resource.',
+            ],
```
this has changed where scopes are dynamically generated. See my PR

adunsulag commented on _rest_config.php:
```diff
@@ -289,6 +289,47 @@ public static function authorization_check($section, $value): void
+                $scope[] = $first . '/*.' . $third;
+                $scope[] = $first . '/' . $second . '.*';
+                $scope[] = $first . '/*.*';
+                $scope[] = '*/' . $second . '.' . $third;
```
So I would recommend we don't allow wildcard scopes.  In the OAUTH2 authorization a client can send us a wildcard scope request but we will send back a specific enumerated list of scopes for the resources we are granting them right @sjpadgett?  So by the time a user makes a request against the REST api they won't be sending us any wildcard scopes because their access token that we generate will have only an enumerated list of specific scope grants.

The specification says if we support wildcard access tokens we have to provide both current access as well as grant access to all FUTURE resources / permissions which I think can be a pretty big security whole if we aren't careful.  That's why I recommend we don't handle wildcards. 

Hopefully I'm making sense there.

adunsulag commented on _rest_config.php:
```diff
@@ -289,6 +289,47 @@ public static function authorization_check($section, $value): void
+                    }
+                }
+                if (!$passScope) {
+                    http_response_code(401);
```
I wasn't sure if just passing 401 was sufficient or if we needed to pass anything else to be spec compliant but this good for FHIR R4.  @see https://www.hl7.org/fhir/security.html#AccessDenied

bradymiller commented on src/Common/Logging/SystemLogger.php:
```diff
@@ -32,7 +32,7 @@ public function __construct()
          */
         $this->logger = new Logger('OpenEMR');
-        $logLevel = Logger::WARNING; // change this if you want to filter what logs you see.
+        $logLevel = Logger::DEBUG; // change this if you want to filter what logs you see.
```
TODO
Note to self. I will need to revert this in future.

bradymiller commented:
hi @sjpadgett and @adunsulag ,

Basically redid this PR:

1. Added several missing scopes (and adjusted the scope autofinding mechanism to find them)
2. Changed autofinding mechanism to derive the patient/ scopes from the patient-specific routes
3. Added permission mechanism in each endpoint to check for scope
4. In dispatch, ensured that had the permission for the requested route (api:oemr, api:fhir, etc.)

Issues:
1. The scope validation mechanism appears to be super slow when sending lots of scopes.
2. Couple php warnings to address.

adunsulag commented:
@bradymiller you  mentioned in your commit message that things are working very slow, is it slow in parsing the access token, doing the permission checks?  Just wondering if there's something I can help with.

sjpadgett commented:
@adunsulag Brady fixed. mostly I was rebuilding the entire permitted scopes array for each scope validation iteration. very slow.

sjpadgett commented:
Am I holding you guys up on anything?
I've been reviewing my ccda generator(all javascript) and looks like I need to write some new leafs for the Blue Button template.
Anyway I haven't done anything on manual registration but will work on locking down dynamic registration for assigning scopes to client tomorrow.

adunsulag commented:
@bradymiller @sjpadgett I looked through what Brady posted and it all looks fine to me, it'd be helpful for me to get this PR in sooner than later, but I'm not held up by anything your doing @sjpadgett.  I have made the one change to the dynamic registration in order to support the launch URI, haven't done anything else beyond that.

bradymiller commented:
Feel free to bring in now if needed (I won't be able to get to github until
late this evening). Just need to turn off the DEBUG setting in the logger,
which I can do to master later this evening (if this PR is brought in
first).

On Mon, Dec 28, 2020 at 11:46 AM Stephen Nielson <notifications@github.com>
wrote:

> @bradymiller <https://github.com/bradymiller> @sjpadgett
> <https://github.com/sjpadgett> I looked through what Brady posted and it
> all looks fine to me, it'd be helpful for me to get this PR in sooner than
> later, but I'm not held up by anything your doing @sjpadgett
> <https://github.com/sjpadgett>. I have made the one change to the dynamic
> registration in order to support the launch URI, haven't done anything else
> beyond that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/4123#issuecomment-751842470>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEDOFZFP3BVLDE3HSW5DLSXDOARANCNFSM4VJRH2OQ>
> .
>

sjpadgett commented:
Okay thanks for this @bradymiller . @adunsulag Merging this now

adunsulag commented:
@bradymiller @sjpadgett I'm putting this here as it relates to this PR.  SMART requires & inferno checks that we send back the list of valid scopes as part of the token response (I originally thought we would embed this inside the id_token as a claim in addition to it already being embedded inside the access_token).  However, our **site:default** is causing the inferno test suite to fail.
![image](https://user-images.githubusercontent.com/228117/103239760-84bfb400-491c-11eb-8974-2975b5bc408e.png)


My thoughts are that in that specific 'scopes' object I'll just remove the 'site:default' and the other stuff so we can be spec compliant.  They will still be embedded in the access_token.

sjpadgett commented:
Having custom scopes is compliant for OIDC. Don't know what inferno prob is but as long as it gets into access token then, well, just then!:)
